z3/src/tactic.rs

```diff
@@ -0,0 +1,159 @@
+use std::ffi::{CStr, CString};
+use std::fmt;
+
+use z3_sys::*;
+use Context;
+use Goal;
+use Params;
+use Tactic;
+use Z3_MUTEX;
+
+impl<'ctx> Tactic<'ctx> {
+    pub fn list_all(ctx: &'ctx Context) {
+        let p = unsafe {
+            Z3_get_num_tactics(ctx.z3_ctx)
+        };
+        for n in 0..p {
+            let t = unsafe {
+                Z3_get_tactic_name(ctx.z3_ctx, n)
+            };
+            match unsafe { CStr::from_ptr(t) }.to_str() {
+                Ok(s) => println!("{}", s),
+                Err(_) => (),
+            }
+        }
+    }
+
+    fn new_from_z3(ctx: &'ctx Context, z3_tactic: Z3_tactic) -> Tactic<'ctx> {
+        Tactic {
+            ctx,
+            z3_tactic,
+        }
+    }
+
+    pub fn new(ctx: &'ctx Context, name: &'ctx str) -> Tactic<'ctx> {
+        let tactic_name = CString::new(name).unwrap();
+        Tactic {
+            ctx,
+            z3_tactic: unsafe {
+                let guard = Z3_MUTEX.lock().unwrap();
+                let t = Z3_mk_tactic(ctx.z3_ctx, tactic_name.as_ptr());
+                Z3_tactic_inc_ref(ctx.z3_ctx, t);
+                t
+            },
+        }
+    }
+
+    /// Return a tactic that just return the given goal.
+    pub fn create_skip(ctx: &'ctx Context) -> Tactic<'ctx> {
+        Tactic::new_from_z3(ctx, unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            let t = Z3_tactic_skip(ctx.z3_ctx);
+            Z3_tactic_inc_ref(ctx.z3_ctx, t);
+            t
+        })
+    }
+
+    /// Return a tactic that always fails.
+    pub fn create_fail(ctx: &'ctx Context) -> Tactic<'ctx> {
+        Tactic::new_from_z3(ctx, unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            let t = Z3_tactic_fail(ctx.z3_ctx);
+            Z3_tactic_inc_ref(ctx.z3_ctx, t);
+            t
+        })
+    }
+
+    /// Return a tactic that keeps applying `t` until the goal is not modified anymore or the maximum
+    /// number of iterations `max` is reached.
+    pub fn repeat(ctx: &'ctx Context, t: Tactic, max: u32) -> Tactic<'ctx> {
+        Tactic {
+            ctx,
+            z3_tactic: unsafe {
+                let guard = Z3_MUTEX.lock().unwrap();
+                let t = Z3_tactic_repeat(ctx.z3_ctx, t.z3_tactic, max);
+                Z3_tactic_inc_ref(ctx.z3_ctx, t);
+                t
+            },
+        }
+    }
+
+    /// Return a tactic that applies the current tactic to a given goal and
+    /// the `then_tactic` to every subgoal produced by the original tactic.
+    pub fn and_then(&self, then_tactic: Tactic) -> Tactic {
+        let t = unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            let t = Z3_tactic_and_then(self.ctx.z3_ctx, self.z3_tactic, then_tactic.z3_tactic);
+            Z3_tactic_inc_ref(self.ctx.z3_ctx, t);
+            t
+        };
+        Tactic {
+            ctx: self.ctx,
+            z3_tactic: t,
```

Comment:
any reason not to keep the unsafe block here?

Reply:
I left that out since it wasn't strictly necessary.  I'm happy to oblige if that's the preference.